numpy/lib/function_base.py

```diff
@@ -1320,6 +1320,258 @@ def gradient(f, *varargs, **kwargs):
     else:
         return outvals
 
+def divergence(v,*varargs,**kwargs):
+    """
+    Return the divergence of an N-dimensional vector field of N 
+    components, each of dimension N.
+
+    The divergence is computed using second order accurate central
+    differences in the interior and either first differences or second
+    order accurate one-sides (forward or backwards) differences at the
+    boundaries. The returned gradient hence has the same shape as the
+    input array.
+    
+    Parameters
+    ----------  
+    v : list of numpy arrays
+        Each of these array is the N-dimensional projection of the vector
+        field on the corresponding axis. 
+    
+    varargs : scalar or list of scalar, optional
+        N scalars specifying the sample distances for each dimension,
+        i.e. `dx`, `dy`, `dz`, ... Default distance: 1.
+        single scalar specifies sample distance for all dimensions.
+        if `axis` is given, the number of varargs must equal the number of axes.
+    edge_order : {1, 2}, optional
+        Gradient is calculated using N\ :sup:`th` order accurate differences
+        at the boundaries. Default: 1.
+        
+            
+    Returns
+    -------
+    divergence : numpy array
+        The output corresponds to the divergence of the input vector field. 
+        This means that the output array has the form 
+        dAx/dx + dAy/dy + dAz/dz + ... for an input vector field 
+        A = (Ax, Ay, Az, ...) up to N dimensions.
+    
+    Example
+    -------
+    This example shows how the calculated divergence of the 2-D field 
+    (0.5*x**2, -y*x) (whose divergence should be 0 everywhere) returns a 0 array
+    
+    >>> import numpy as np
+    >>> X,Y=np.mgrid[0:2000,0:2000]
+    >>> a1=0.5*X**2
+    >>> a2=-Y*X
+    >>> c=[a1,a2]
+    >>> d=np.divergence(c)
+    >>> print d
+    [[ 0.  0.  0. ...,  0.  0.  0.]
+     [ 0.  0.  0. ...,  0.  0.  0.]
+     [ 0.  0.  0. ...,  0.  0.  0.]
+     ..., 
+     [ 0.  0.  0. ...,  0.  0.  0.]
+     [ 0.  0.  0. ...,  0.  0.  0.]
+     [ 0.  0.  0. ...,  0.  0.  0.]
+    """
+    N = [0]*len(v)
+    for i, v_c in enumerate(v):
+        v_c = np.asanyarray(v_c)
+        N[i] = len(v_c.shape) # Get number of dimensions from every component
+    if False in [N[0] == N[i] for i in range(len(v))]: 
+        #Check if all components are the same size
+        raise ValueError("Not all components of input"
+	                 " have the same number of dimensions")
+    else:
+        if False in [np.shape(v[0]) == np.shape(v[i]) for i in range(len(v))]:
+            raise ValueError("Not all components of input are the same size")
+        else:
+	    N = N[0] 
+	# If all vector field components are the same shape,
+	#N becomes the number of dimensions
+    
+    axes = tuple(range(N))
```

Comment:
Actually, in this case we want the result to be added in-place on the first object to avoid an unnecessary memory allocation. So really, we need something like this:
```python
def inplace_sum(items):
    it = iter(items)
    total = next(it)
    for item in it:
        total += item
    return total
```